File: app/src/main/java/com/alternativeinfrastructures/noise/models/RemoteIdentity.java
package com.alternativeinfrastructures.noise.models;

import com.alternativeinfrastructures.noise.NoiseDatabase;
import com.alternativeinfrastructures.noise.models.signal.TypeConverters;
import com.dbflow5.annotation.Column;
import com.dbflow5.annotation.Index;
import com.dbflow5.annotation.PrimaryKey;
import com.dbflow5.annotation.Table;
import com.dbflow5.reactivestreams.structure.BaseRXModel;

import org.whispersystems.libsignal.IdentityKey;

@Table(database = NoiseDatabase.class)
public class RemoteIdentity extends BaseRXModel {
    public RemoteIdentity(String username, int deviceId, IdentityKey identityKey) {
        this.username = username;
        this.deviceId = deviceId;
        this.identityKey = identityKey;
    }

    public RemoteIdentity() {}

    @Override
    public boolean equals(Object object) {
        RemoteIdentity other;
        if (!(object instanceof RemoteIdentity))
            return false;

        other = (RemoteIdentity) object;
        return (username.equals(other.username) &&
                deviceId == other.deviceId &&
                identityKey.equals(other.identityKey));
    }

    @Column
    @Index
    protected String username;
    public String getUsername() { return username; }

    @Column
    protected int deviceId;
    public int getDeviceId() { return deviceId; }

    @PrimaryKey
    @Column(typeConverter = TypeConverters.IdentityKeyConverter.class)
    protected IdentityKey identityKey;
    public IdentityKey getIdentityKey() { return identityKey; }
}

File: app/src/main/java/com/alternativeinfrastructures/noise/models/RemoteIdentity.kt
package com.alternativeinfrastructures.noise.models

import com.alternativeinfrastructures.noise.NoiseDatabase
import com.alternativeinfrastructures.noise.models.signal.TypeConverters
import com.raizlabs.android.dbflow.annotation.Column
import com.raizlabs.android.dbflow.annotation.Index
import com.raizlabs.android.dbflow.annotation.PrimaryKey
import com.raizlabs.android.dbflow.annotation.Table
import com.raizlabs.android.dbflow.rx2.structure.BaseRXModel

import org.whispersystems.libsignal.IdentityKey

@Table(database = NoiseDatabase::class)
class RemoteIdentity : BaseRXModel {

    @Column
    @Index
    var username = ""

    @Column
    var deviceId: Int = 0

    @PrimaryKey
    @Column(typeConverter = TypeConverters.IdentityKeyConverter::class)
    lateinit var identityKey: IdentityKey

    constructor(username: String, deviceId: Int, identityKey: IdentityKey) {
        this.username = username
        this.deviceId = deviceId
        this.identityKey = identityKey
    }

    constructor() {}

    override fun equals(`object`: Any?): Boolean {
        if (`object` !is RemoteIdentity)
            return false

        return username == `object`.username &&
                deviceId == `object`.deviceId &&
                identityKey == `object`.identityKey
    }
}
